src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -499,12 +500,63 @@ void StorageObjectStorage::truncate(
             getName(), path.path);
     }
 
-    StoredObjects objects;
-    for (const auto & key : configuration->getPaths())
+    if (configuration->partition_strategy && configuration->partition_strategy_type == PartitionStrategyFactory::StrategyType::HIVE) 
+    {
+        auto query_settings = configuration->getQuerySettings(context);
+        query_settings.throw_on_zero_files_match = false;
+        query_settings.ignore_non_existent_file = true;
+    
+        bool local_distributed_processing = distributed_processing;
+        if (context->getSettingsRef()[Setting::use_hive_partitioning])
+            local_distributed_processing = false;
+    
+        auto iterator = StorageObjectStorageSource::createFileIterator(
```

Comment:
I think this code will actually work for both cases, maybe there is no need for two truncate algorithms.

For hive partition strategy, `StorageObjectStorageSource::createFileIterator` will use the glob iterator. Basically iterating over all files that a `read` operation would see.

It`ll do the same for the non hive case (i.e, loop over the `paths` object). See https://github.com/arthurpassos/ClickHouse/blob/master/src/Storages/ObjectStorage/StorageObjectStorageSource.cpp#L247

Reply:
Can you add non hive partitioned tests?

Reply:
Done.